README.md
# CSmosquitoSp
## R package for Spatial analysis of Citizen Science mosquito reports

The package provides a collection of analytics to perform spatially-oriented analysis of Citizen Science mosquito reports for specific Area Of Interest and defined temporal interval. It provides function to handle spatial data and to generate map layout.
The package allows to import spatialized mosquito reports (at present ZanzaMapp DataBase is supported), perform spatio-temporal analysis of the report dataset, create spatial map layoutsand export results.
Spatial polygons for the Italian territory are provided. Users can upload custom spatial polygons 

### Main features:

* import spatialized mosquito reports (at present ZanzaMapp DataBase is supported)
* clean up reports
* perform spatial analysis of the report dataset for a defined temporal interval
* create spatial map layouts
* export results

### Input spatial features:

* Vector spatial features of class 'sf' with the following fields: 'CODE' (unique code); 'AREA' (in square km); 'POPULATION'; 'POP_DENS' (population every 1 square km)

### Output:

* Spatial vector features in Shapefile format (and optionally KML format) containing results computed for each spatial polygon
* Map layout in PNG format

### Installation

**To load** (using `devtools`):
```r
library(devtools)
install_github("ffilipponi/CSmosquitoSp")
```
### Use in QGIS

The package can be used from QGIS. In QGIS version 3 it is necessary to install 'Processing R Provider' plugin. The CSmosquitoSp spatial analyst can be integrated in QGIS Processing by adding the files /inst/rscript/CSmosquitoSp_spatial_analyst.rsx and /inst/rscript/CSmosquitoSp_spatial_analyst.rsx in the R script folder used by QGIS R processing.

### Authors

* Filipponi Federico
* Manica Mattia

### License

Licensed under the GNU General Public License, Version 3.0: https://www.gnu.org/licenses/gpl-3.0.html


ChangeLog
--- to do


--- current version

2020-07-01 Federico Filipponi <federico.filipponi@gmail.com>

* Releasing the fifth version of the package (version 0.5)
* Define final package name to 'CSmosquitoSp'
* Added new features to the analysis function
* Edited plot function for compatibility with newer 'ggplot2' and 'sf' packages
* Minor bug fixes

2018-03-19 Federico Filipponi <federico.filipponi@gmail.com>

* Releasing the forth version of the package (version 0.4)
* Increased support for the use of custom external vector spatial objects
* Minor bug fixes


2018-03-13 Federico Filipponi <federico.filipponi@gmail.com>

* Releasing the third version of the package (version 0.3)
* Changed package for handling vector spatial objects from 'sp' to 'sf'
* Changed package to generate maps to 'ggplot2'
* Moved spatial join function from 'spatialEco::point.in.poly()' to 'sf::st_intersection()'
* Created QGis R script to execute data analysis from QuantumGIS software
* Minor bug fixes


2017-11-23 Federico Filipponi <federico.filipponi@gmail.com>

* Releasing the second version of the package (version 0.2)
* Minor bug fixes

2017-10-06 Federico Filipponi <federico.filipponi@gmail.com>

* Releasing the first version of the package (version 0.1)
* Starting this log


R/CSmosquitoSp_export.R
# Title         : CSmosquitoSp export
# Description   : Export results from spatial analysis of CSmosquitoSp dataset for a specific Area Of Interest and a selected time interval
# Date          : Jul 2020
# Version       : 0.5
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>

# ######################################################################
#' @title CSmosquitoSp export
#'
#' @description This function export results computed using \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} function
#' a specific Area Of Interest and a selected temporal range.
#'
#' @param x Input object of class \code{\linkS4class{CSmosquitoSp.obj}}
#' @param output_path Character. Path where maps are saved on disk.
#' @param points Logical. When set to TRUE export points used for the analysis also in 'shapefile' format (default to FALSE).
#' @param kml Logical. When set to TRUE results are also exported in 'kml' format (default to FALSE).
#' @param sensitive Logical. When set to TRUE export the sensitive information in records (default to FALSE).
#'
#' @return
#' The function export the results generated using function \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} to disk.
#' Output vector file format are 'shapefile' and 'kml' (optional).
#'
#' @author Federico Filipponi
#'
#' @keywords CSmosquitoSp export
#'
#' @seealso \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} \code{\link[CSmosquitoSp]{CSmosquitoSp.plot}}
#'
#' @examples
#' \dontrun{
#' # perform CSmosquitoSp spatial analysis
#' for the metropolitan area of the city of Rome
#' CSMS_roma <- CSmosquitoSp.analysis(input_db=input_db, aux_path=aux_path,
#' aoi="Roma", start_date="2017-09-01", end_date="2017-09-01")
#'
#' # export result
#' CSmosquitoSp.export(x=CSMS_result, output_path="C:/output")
#'
#' # export result also in kml format
#' CSmosquitoSp.export(x=CSMS_result, output_path="C:/output", kml=TRUE)
#'
#' # export also points used for the analysis
#' CSmosquitoSp.export(x=CSMS_result, output_path="C:/output", points=TRUE)
#' }
#'
#' @import sp
#' @import sf
#' @import methods
#' @importFrom utils write.table
#' @importFrom sf st_write
#' @importFrom rgdal writeOGR
#' @importFrom methods new
#' @importFrom methods as
#' @export

# define 'CSmosquitoSp.export' function
CSmosquitoSp.export <- function(x, output_path, points=FALSE, kml=FALSE, sensitive=FALSE){

  # ##################################
  # check input arguments

  # check if output path exists
  if(!is.null(output_path)){
    output_path <- normalizePath(output_path, winslash="/", mustWork=FALSE)
    if(!file.exists(output_path)){
      dir.create(output_path, recursive = T)
    }
  }

  # check if 'x' argument is of class 'CSmosquitoSp.obj'
  if(!(class(x) %in% c("CSmosquitoSp.obj")))
    stop("'x' argument must be one an object of class 'CSmosquitoSp.obj'\ngenerated using function CSmosquitoSp.analysis()")

  # set option to avoid scientific notation
  options(scipen = 20)

  # load data from the package
  # data(aoi_extent)
  aoi_extent <- aoi_extent

  if(!(x@aoi_name %in% aoi_extent$aoi)){
    aoi_fname <- x@aoi_name
  } else {
    aoi_fname <- aoi_extent$aoi[which(aoi_extent$aoi == x@aoi_name)]
  }

  # set basename for output files
  out_basename <- paste(aoi_fname, "_", as.character(x@start_date), "_", as.character(x@end_date), sep="")

  # ####################################
  # Remove sensitive information

  if(!sensitive){
    x@points$deviceid <- NULL
    x@points$devicetype <- NULL
  } else {
    class(x@points$deviceid) <- "character"
  }

  # ####################################
  # Export data

  # to CSV file
  # convert to data frame
  export_df <- st_set_geometry(x@aoi, NULL)
  write.table(export_df, file=normalizePath(path=paste(output_path, "/", out_basename, ".dataset.csv", sep=""), winslash="/", mustWork=F), sep="\t", dec=".", quote=FALSE, col.names=TRUE, row.names=FALSE, na="", fileEncoding="UTF-8")

  # to shapefile
  # # using 'sp' package
  # rgdal::writeOGR(x@aoi, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".data.shp", sep=""), winslash="/", mustWork=F), layer=aoi, driver="ESRI Shapefile")
  # using 'sf' package
  sf::st_write(x@aoi, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".data.shp", sep=""), winslash="/", mustWork=F), driver="ESRI Shapefile", factorsAsCharacter=TRUE, delete_layer=TRUE, quiet=TRUE)

  # write fields description
  # result_legend <- CSmosquitoSp:::result_legend
  write.table(result_legend, file=normalizePath(path=paste(output_path, "/", out_basename, ".dataset_fields", ".csv", sep=""), winslash="/", mustWork=F), sep="\t", dec=".", quote=FALSE, col.names=TRUE, row.names=FALSE)

  # export analysis points
  if(points){
    # convert field types to character in order to avoid warnigs related to long numeric format
    class(x@points$posix) <- "character"
    class(x@points$date) <- "character"

    # save cleaned and subsetted CSmosquitoSp DataBase to CSV
    write.table(x@points, file=normalizePath(path=paste(output_path, "/", out_basename, ".points", ".csv", sep=""), winslash="/", mustWork=F), sep="\t", dec=".", quote=FALSE, col.names=TRUE, row.names=FALSE, na="", fileEncoding="UTF-8")
    # to shapefile
    # # using 'sp' package
    # rgdal::writeOGR(x@points, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".points.shp", sep=""), winslash="/", mustWork=F), layer="points", driver="ESRI Shapefile")
    # using 'sf' package
    sf::st_write(x@points, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".points.shp", sep=""), winslash="/", mustWork=F), driver="ESRI Shapefile", factorsAsCharacter=TRUE, delete_layer=TRUE, quiet=TRUE)

    # export to kml format
    if(kml){
      # to kml file
      # convert from 'sf' to 'Spatial' (sp) object
      export_aoi <- as(x@aoi, "Spatial")
      # export using 'sp' package
      rgdal::writeOGR(export_aoi, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".data.kml", sep=""), winslash="/", mustWork=F), layer=aoi, driver="KML", dataset_options=list("NameField=CODE", "DescriptionField=VAL_PMFA", "AltitudeMode=clampToGround"), overwrite_layer=TRUE)
      # to kml file
      # convert from 'sf' to 'Spatial' (sp) object
      export_points <- as(x@points, "Spatial")
      # export using 'sp' package
      rgdal::writeOGR(export_points, dsn=normalizePath(path=paste(output_path, "/", out_basename, ".points.kml", sep=""), winslash="/", mustWork=F), layer=aoi, driver="KML", dataset_options=list("NameField=CODE", "DescriptionField=VALUE", "AltitudeMode=clampToGround"), overwrite_layer=TRUE)

    }
  }
}


R/CSmosquitoSp_import.R
# Title         : CSmosquitoSp import
# Description   : Import CSmosquitoSp DataBase
# Date          : Jul 2020
# Version       : 0.5
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>

# ######################################################################
#' @title CSmosquitoSp import
#'
#' @description This function import CSmosquitoSp DataBase to be used for spatial analysis using function \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}}
#'
#' @param x Character. Input path where CSmosquitoSp DataBase in CSV format is stored on disk.
#' @param type Character. Type of the imported Citizen Science mosquito DataBase (at present only 'ZanzaMapp' type is supported)
#'
#' @return
#' The function return a DataFrame containing the CSmosquitoSp DataBase
#'
#' @author Federico Filipponi
#'
#' @keywords CSmosquitoSp import
#'
#' @seealso \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} \code{\link[CSmosquitoSp]{CSmosquitoSp.plot}} \code{\link[CSmosquitoSp]{CSmosquitoSp.export}}
#'
#' @examples
#' \dontrun{
#' # set input path
#' input_path <- normalizePath(path=system.file("external/CSmosquitoSp_database.csv"),
#' winslash="/", mustWork=TRUE)
#' # import mosquito DataBase of type 'ZanzaMapp'
#' input_db <- CSmosquitoSp.import(x=input_path, type="ZanzaMapp")
#'
#' ###
#' # directly input mosquito DataBase file path in the function
#' input_db <- CSmosquitoSp.import(x="C:/CSmosquitoSp_database.csv", type="ZanzaMapp")
#' }
#' @importFrom utils read.table
#' @export

# define 'CSmosquitoSp.plot' function
CSmosquitoSp.import <- function(x, type=NULL){

  # ##################################
  # check input arguments

  # check if input path exists
  if(!file.exists(x))
    stop("Input file does not exist on disk")
  
  if(is.null(type))
    stop("Input file type not defined using argument 'type'")
  
  if(type != "ZanzaMapp")
    stop("At present only 'ZanzaMapp' mosquito DataBase is supported")
  
  # import ZanzaMapp DataBase
  if(type == "ZanzaMapp"){
    # check if input file is a standard ZanzaMapp DataBase
    if(readLines(con=x, n=1) != c("locationid,ts,lat,lng,val,deviceid,tilex,tiley,time,place,bites,devicelat,devicelon,devicetype"))
      stop("Input file is not a standard ZanzaMapp DataBase")
    
    # import ZanzaMapp DB
    CSmosquitoSp_db <- read.table(file=normalizePath(path=x, winslash="/", mustWork=T), header=T, sep=",", dec=".")
    comment(CSmosquitoSp_db) <- "ZanzaMapp"
    
    return(CSmosquitoSp_db)
  }
}


R/CSmosquitoSp_classes.R
# Title         : Classes for the CSmosquitoSp package
# Description   : Define classes for the CSmosquitoSp package
# Date          : Jul 2020
# Version       : 0.5
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>

# ######################################################################
#
#' Class "CSmosquitoSp.obj"
#'
#' @name CSmosquitoSp.obj
#' @aliases CSmosquitoSp.obj-class
#' @description Result from \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} stored as a \code{\linkS4class{CSmosquitoSp.obj}} object.
#' Slots for \code{CSmosquitoSp.obj} objects include: (1) the reported points from the mosquito DataBase;
#' (2) the AOI spatial polygons containing results from the spatial analysis of the mosquito DataBase.
#' @docType class
#' @rdname CSmosquitoSp-class
#' @section Objects from the class: Objects are created by calls to \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}}.
#'
#' @seealso \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} \code{\link[CSmosquitoSp]{CSmosquitoSp.plot}} \code{\link[CSmosquitoSp]{CSmosquitoSp.export}}
#'
#' @slot DB Character. Type of the imported Citizen Science mosquito DataBase
#' @slot points SpatialPointsDataFrame. Report spatial points used for the analysis
#' @slot aoi SpatialPolygonsDataFrame. Spatial polygons of the selected Area Of Interest containing the results from the spatial analysis
#' @slot aoi_name Character. Name of the selected Area Of Interest
#' @slot start_date Date. Start day selected for the analysis
#' @slot end_date Date. End day selected for the analysis
#' @slot method Character. Method used for the analysis (currently 'RC' and 'CM' are supported)
#' 
#' @author Federico Filipponi
#'
#' @exportClass CSmosquitoSp.obj

setClass("CSmosquitoSp.obj",
         slots=c(DB="character",
                        points="sf",
                        aoi="sf",
                        aoi_name="character",
                        start_date="Date",
                        end_date="Date",
                        method="character"),
                        prototype=prototype(DB=NULL, points=NULL, aoi=NULL, aoi_name=NULL, start_date=NULL, end_date=NULL, method=NULL)
         )


R/CSmosquitoSp_methods.R
# Title         : Methods for the CSmosquitoSp package
# Description   : Define methods for the CSmosquitoSp package
# Date          : Jul 2020
# Version       : 0.5
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>

# ######################################################################

#' @title CSmosquitoSp methods
#' @description This function define methods for the CSmosquitoSp package for class \code{\linkS4class{CSmosquitoSp.obj}}
#' a specific Area Of Interest and a selected temporal range.
#' 
#' @name show
#' @aliases show show,CSmosquitoSp.obj-method
#' @rdname CSmosquitoSp_methods
#' 
#' @param object CSmosquitoSp.obj. Object of class \code{\linkS4class{CSmosquitoSp.obj}}
#' 
#' @author Federico Filipponi
#' 
#' @importFrom utils data
#' @importFrom utils str
#' @exportMethod show

# create function to show object of class 'CSmosquitoSp.obj'
setMethod('show', signature(object='CSmosquitoSp.obj'),
          function(object){
            cat("CSmosquitoSp object of class 'CSmosquitoSp.obj'", '\n')
            cat('\n')
            cat('DataBase type:',  as.character(object@DB), '\n')
            cat('\n')
            cat('Number of spatial points:',  nrow(object@points), '\n')
            cat('Start date:\t', as.character(object@start_date), '\n')
            cat('End date:\t', as.character(object@end_date), '\n')
            cat('Method:\t', as.character(object@method), '\n')
            cat('\n')
            cat('Spatial polygon information', '\n')
            cat('\n')
            cat('Number of polygons:', length(object@aoi), '\n')
            cat('AOI:\t\t', object@aoi_name,'\n')
            cat('Spatial Bounding Box:', '\n')
            cat(show(st_bbox(object@aoi)))
            cat(show(st_crs(object@aoi)), '\n')
            cat('\n')
            cat('Spatial polygon data.frame has',  nrow(object@aoi), 'records', '\n')
            cat(str(object@aoi), '\n')
          }
)


inst/rscript/CSmosquitoSp_spatial_analyst_v0.5.0.R
#!/usr/bin/Rscript
# Title         : CSmosquitoSp spatial analyst
# Description   : Perform spatially-oriented analysis of Citizen Scientists mosquito reports for specific Area Of Interest and a selected time interval
# Date          : Jul 2020
# Version       : 0.5.0
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>
# ######################################################################

# ##################################
# Load required libraries
library(CSmosquitoSp)

# ###########################################################
# set variables
# ###########################################################

# set input ZanzaMapp DataBase file path
input_path <- normalizePath(path="C:/zanzamapp/DB_zanzamApp_20180228.csv", winslash="/", mustWork=T)

# set output directory
output_path <- normalizePath(path="C:/zanzamapp/", winslash="/", mustWork=FALSE)

# set Area Of Interest (currently supported: 'Roma', 'Latina', 'Lazio', 'Italia', 'Italia_comuni', 'Roma_municipi', 'Roma_griglia_1000m', 'Roma_griglia_250m', 'Roma_griglia_100m')
aoi <- c("Lazio")

# set start date
start_date <- as.Date("2017-09-01")

# set end date
end_date <- as.Date("2017-09-30")

# ##################################
# Analyse data

# import ZanzaMapp DataBase
CSMS_db <- CSmosquitoSp.import(input_path, type="ZanzaMapp")

# perform spatial analysis
CSMS_result <- CSmosquitoSp.analysis(input_db=CSMS_db, aoi=aoi, start_date=start_date, end_date=end_date)

# or as alternative set parameters directly in the function
CSMS_result_roma_201707 <- CSmosquitoSp.analysis(input_db=CSMS_db, aoi="Roma", start_date=as.Date("2017-07-01"), end_date=as.Date("2017-07-31"))

# # alternatively load an external polygon vector layer
# library(sf)
# area <- sf::st_read(dsn="/media/DATA/zanzamapp/auxiliary_files/Lazio.shp")
# # perform spatial analysis using the external polygon
# CSMS_result <- CSmosquitoSp.analysis(input_db=CSMS_db, aoi=area, start_date=start_date, end_date=end_date, aoi_name="Area_name")

# ##################################
# Export results

# save results from CSmosquitoSp spatial analysis to disk
CSmosquitoSp.export(x=CSMS_result, output_path=output_path)

# save all results from CSmosquitoSp spatial analysis to disk (included points and export to KML format)
# CSmosquitoSp.export(x=CSMS_result, output_path=output_path, points=TRUE, kml=TRUE)

# ##################################
# Generate maps

# plot map of perceived mosquito abundance
CSmosquitoSp.plot(x=CSMS_result, field="VAL_PMFA")
# plot map of sampling effort
CSmosquitoSp.plot(x=CSMS_result, field="REP_DC_POP")

# plot map of Weighted average of DayTime mosquito (i.e. Aedes albopictus) with Italian title and footer and with caption figure
CSmosquitoSp.plot(x=CSMS_result, field="DT_V_AVG", maptitle="Perceived Aedes albopictus abundance", lang = "ITA", caption_logo = TRUE)

# plot map of population (any numeric field of 'names(CSMS_result$aoi)' is supported)
CSmosquitoSp.plot(x=CSMS_result, field="POPULATION")

# save plot to PNG file
plot_field <- "VAL_AVG"
png_filename <- normalizePath(path=paste(output_path, "/", 
                                         paste(CSmosquitoSp:::aoi_extent$aoi[which(CSmosquitoSp:::aoi_extent$aoi == CSMS_result@aoi_name)], "_", as.character(CSMS_result@start_date), "_", as.character(CSMS_result@end_date), sep=""),
                                         "_map_", plot_field, ".png", sep=""), winslash="/", mustWork=FALSE)
CSmosquitoSp.plot(x=CSMS_result, field=plot_field, filename=png_filename)

# ##################################
# Display results in R

# show CSmosquitoSp spatial analysis results
CSMS_result

# get table of spatial points used for the analysis
CSMS_points <- st_set_geometry(CSMS_result@points, NULL)

# get table with the results from the spatial analysis
CSMS_table <- st_set_geometry(CSMS_result@aoi, NULL)

# display the descrition of the resulting field names
CSmosquitoSp:::result_legend

# ##################################
# Development section

# perform spatial analysis using the DB 'full_clean' option
CSMS_result <- CSmosquitoSp.analysis(input_db=CSMS_db, aoi=aoi, start_date=start_date, end_date=end_date, full_clean=TRUE)

# you can edit the source code of 'CSmosquitoSp.analysis' to modify how the values are calculated, and load it as a source
source("C:/CSmosquitoSp_analysis.R")
# perform spatial analysis
CSMS_result <- CSmosquitoSp.analysis(input_db=CSMS_db, aoi=aoi, start_date=start_date, end_date=end_date)


R/CSmosquitoSp_plot.R
# Title         : CSmosquitoSp plot
# Description   : Create maps from spatial analysis results of CSmosquitoSp dataset for a specific Area Of Interest
# Date          : Jul 2020
# Version       : 0.5
# Licence       : GPL v3
# Authors       : Federico Filipponi
# Maintainer    : Federico Filipponi <federico.filipponi@gmail.com>

# ######################################################################
#' @title CSmosquitoSp plot
#'
#' @description This function create spatial maps for a selected AOI using the results
#' generated from spatial analysis of CSmosquitoSp data
#'
#' @param x Input object of class \code{\linkS4class{CSmosquitoSp.obj}}
#' @param field Character. Fields to be mapped from the 'aoi' slot of object of class \code{\linkS4class{CSmosquitoSp.obj}}
#' @param maptitle Character. 'Map title.
#' @param lang Character. Language to use for map footer ('ENG' or 'ITA').
#' @param filename Character. Plot result to speified file path.
#' @param caption_logo Logical. Add the caption logo (currently available only for 'ZanzaMapp' DataBase) to the map (Default to FALSE).
#' @param ... Additional arguments to be passed through to function \code{\link{CSmosquitoSp.plot}}
#'
#' @return
#' The function return a plot for the selected field computed from \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} within the Area Of Interest
#'
#' @author Federico Filipponi
#'
#' @keywords CSmosquitoSp plot
#'
#' @seealso \code{\link[CSmosquitoSp]{CSmosquitoSp.analysis}} \code{\link[CSmosquitoSp]{CSmosquitoSp.export}}
#'
#' @examples
#' \dontrun{
#' # perform CSmosquitoSp spatial analysis
#' for the metropolitan area of the city of Rome
#' CSMS_roma <- CSmosquitoSp.analysis(input_db=input_db, aux_path=aux_path,
#' aoi="Roma", start_date="2017-09-01", end_date="2017-09-30")
#'
#' # plot result
#' CSmosquitoSp.plot(x=CSMS_roma, field="VAL_PMFA")
#' }
#'
#' @import sf
#' @import methods
#' @import ggplot2
#' @import ggpubr
#' @importFrom grDevices dev.off
#' @importFrom grDevices png
#' @importFrom ggplot2 ggplot
#' @importFrom ggplot2 theme
#' @importFrom ggplot2 labs
#' @importFrom ggplot2 element_text
#' @importFrom ggplot2 element_line
#' @importFrom ggplot2 element_rect
#' @importFrom ggplot2 xlim
#' @importFrom ggplot2 ylim
#' @importFrom ggplot2 geom_sf
#' @importFrom ggplot2 scale_fill_manual
#' @importFrom ggplot2 scale_fill_distiller
#' @importFrom ggpubr ggarrange
#' @export

# define 'CSmosquitoSp.plot' function
CSmosquitoSp.plot <- function(x, field="VAL_PMFA", maptitle="Perceived mosquito abundance", lang="ENG", filename=NULL, caption_logo=FALSE, ...){

  # ##################################
  # check input arguments

  # log_trans = FALSE
  
  # check if output path exists
  if(!is.null(filename)){
    filename <- normalizePath(filename, winslash="/", mustWork=FALSE)
    if(!file.exists(dirname(filename))){
      dir.create(dirname(filename))
    }
  }

  # check if 'x' argument is of class 'CSmosquitoSp.obj'
  if(!(class(x) %in% c("CSmosquitoSp.obj")))
    stop("'x' argument must be one an object of class 'CSmosquitoSp.obj'\ngenerated using function CSmosquitoSp.analysis()")

  # check if 'field' argument is supported
  if(!(field %in% names(x@aoi)))
    stop("'field' argument must be a field name of the 'aoi' slot in object 'CSmosquitoSp.obj'")
  # if(!(field %in% c("VAL_PMFA", "REP_DC_POP")))
  #   warning("'field' argument should be one of the following in order to generate standard plots:\n* 'VAL_AVG'\n* 'DT_V_AVG'")
  
  # load data from the package
  # data(aoi_extent)
  # aoi_extent <- CSmosquitoSp:::aoi_extent
  
  # set external-aoi variable
  external_aoi <- FALSE
  
  # set variables for external aoi
  if(!(x@aoi_name %in% aoi_extent$aoi)){
    external_aoi <- TRUE
    aoi_extent <- as.data.frame(rbind(as.matrix(aoi_extent), c(x@aoi_name, x@aoi_name, "aoi_poly", x@aoi_name, x@aoi_name, as.double(st_bbox(x@aoi)[1]), as.double(st_bbox(x@aoi)[3]), as.double(st_bbox(x@aoi)[2]), as.double(st_bbox(x@aoi)[4]), as.double(0.05), as.double(0.05))))
  }

  # ##################################
  # get lines for overlay
  aoi_line_name <- as.character(aoi_extent$contour_filename[which(aoi_extent$aoi == x@aoi_name)])
  if(external_aoi){
    aoi_line <- x@aoi
  } else {  
    aoi_line <- get(aoi_line_name)
  }

  # # set default column number
  # df_colnum <- 21
  # set column number
  df_colnum <- as.integer(which(names(x@aoi) == field))

  # ##################################
  # set parameters based on selected 'field'
  
  if(!(field %in% names(x@aoi))){
    stop("'field' argument must be a field name of the 'aoi' slot in object 'CSmosquitoSp.obj'")
  } else {
    if(!is.numeric(as.data.frame(x@aoi[,df_colnum])[,1]))
      stop("'field' argument must be a numeric field of the 'aoi' slot in object 'CSmosquitoSp.obj'")
    
    # set color palette
    scale_palette <- "YlOrRd"
    
    # convert numeric values to class values for plotting
    x@aoi$Legenda <- as.numeric(as.data.frame(x@aoi)[,df_colnum])
    # names(x@aoi)[which(names(x@aoi) == "Legenda")] <- field ### il nome viene duplicato
  }
  if(field == "VAL_PMFA"){
    # set title and color rules
    maptitle <- c("Perceived mosquito abundance")
    scale_palette <- "YlOrRd"
  }
  if(field == "REP_DC_POP"){
    # set title and color rules
    maptitle <- c("Sampling effort")
    scale_palette <- "Greens"
    # log_trans = TRUE
    # rescaled_values <- c(0,0.00000001,0.0000001,0.000001,0.00001,0.0001)
  }
  
  # # set log transform
  # if(log_trans == TRUE){
  #   # x@aoi[,df_colnum] <- as.vector(log10(st_drop_geometry(x@aoi)[,df_colnum]))
  #   field_perc <- as.vector(quantile(st_drop_geometry(x@aoi)[,df_colnum], probs = c(0.01, 0.99), na.rm=TRUE))
  #   rescaled_values <- scales::rescale(10^(seq(field_perc[1], field_perc[2], length.out = 6)))
  # }

  # #######
  # set plot parameters

  # set plot footer
  if(lang == "ITA"){
    ref_period_text <- "Periodo di riferimento"
    if(maptitle == "Perceived mosquito abundance"){
      maptitle <- "Abbondanza di zanzare percepito"
    }
  }
  if(lang == "ENG"){
    ref_period_text <- "Reference period"
  }
  
  # set plot subtitles
  map_subtitle <- as.character(aoi_extent$description[which(aoi_extent$aoi == x@aoi_name)])
  caption_text <- paste(ref_period_text , ": ", sprintf("%02i", as.numeric(strftime(x@start_date,format = "%d" ))), "/", sprintf("%02i", as.numeric(strftime(x@start_date,format = "%m" ))), "/", as.numeric(strftime(x@start_date,format = "%Y" )),
                   " - ", sprintf("%02i", as.numeric(strftime(x@end_date,format = "%d" ))), "/", sprintf("%02i", as.numeric(strftime(x@end_date,format = "%m" ))), "/", as.numeric(strftime(x@end_date,format = "%Y" )), sep="")
  polygon_lwd <- as.double(as.character(aoi_extent$lwd[which(aoi_extent$aoi == x@aoi_name)]))
  contour_lwd <- as.double(as.character(aoi_extent$lwd_overlay[which(aoi_extent$aoi == x@aoi_name)]))

  # #######
  # create plot object
  zm_plot <- ggplot2::ggplot() +
    # mapTheme() +
    ggplot2::theme(
      text = ggplot2::element_text(color = "black"),
      plot.title = ggplot2::element_text(size = 18, color = "black"),
      plot.subtitle = ggplot2::element_text(size = 14, color = "black"),
      plot.caption = ggplot2::element_text(face="italic", size=10),
      axis.text.x = ggplot2::element_text(size=10),
      axis.text.y = ggplot2::element_text(angle=90, size=10),
      panel.background = ggplot2::element_rect(fill="transparent", color="black"),
      panel.grid.major = ggplot2::element_line(size=0.5, color="#343434", linetype=3),
      panel.border = ggplot2::element_rect(color = "black", fill=NA, size=1.0),
      panel.ontop = TRUE
    ) +
    ggplot2::xlim(c(as.double(as.character(aoi_extent$xmin[which(aoi_extent$aoi == x@aoi_name)])), as.double(as.character(aoi_extent$xmax[which(aoi_extent$aoi == x@aoi_name)])))) +
    ggplot2::ylim(c(as.double(as.character(aoi_extent$ymin[which(aoi_extent$aoi == x@aoi_name)])), as.double(as.character(aoi_extent$ymax[which(aoi_extent$aoi == x@aoi_name)])))) +
    ggplot2::geom_sf(data=x@aoi, aes(fill=Legenda), color="black", size=polygon_lwd) +
    # scale_fill_manual(values=color_rules, drop=FALSE) +
    ggplot2::geom_sf(data=aoi_line, fill=NA, color="black", size=contour_lwd) +
    ggplot2::labs(title=maptitle, subtitle=map_subtitle, caption=caption_text)

  # apply color ramp
  zm_plot <- zm_plot + ggplot2::labs(fill=field) + ggplot2::scale_fill_distiller(palette=scale_palette, direction=1, na.value="#f2f2f2")
  # zm_plot <- zm_plot + ggplot2::labs(fill=field) + ggplot2::scale_fill_distiller(palette=scale_palette, direction=1, values=rescaled_values, na.value="#f2f2f2")
  
  # add caption logo image
  if(caption_logo){
    if(x@DB == "ZanzaMapp"){
      # get map caption image
      map_caption <- get("map_caption")
      # plot with caption image
      map_caption_plot <- plot(map_caption) + ggpubr::bgcolor("white")
      z_plot <- ggpubr::ggarrange(zm_plot, map_caption_plot, ncol=1, nrow=2, heights=c(1, 0.3))
    } else {
      z_plot <- zm_plot
    }
  } else {
    z_plot <- zm_plot
  }

  # generate plot
  if(!is.null(filename)){
    png(file=filename, units="px", res=300, width=1800, height=1800, bg="white", pointsize=6)
    # # using 'sp' package
    # print(sp::spplot(x@aoi, zcol=field, main=paste(aoi_extent$description[which(aoi_extent$aoi == x@aoi_name)], "\nNumero medio di zanzare", sep=""), sub=sub_title,
    #              col.regions=zanza.palette, cuts=5, at=breaks, scales=list(draw=TRUE, tck=c(1,0)), col="grey26", lwd=0.5))
    # using 'sf' package
    print(z_plot)
    dev.off()
  } else {
    # # using 'sp' package
    # print(sp::spplot(x@aoi, zcol=field, main=paste(aoi_extent$description[which(aoi_extent$aoi == x@aoi_name)], "\nNumero medio di zanzare", sep=""), sub=sub_title,
    #                    col.regions=zanza.palette, cuts=5, at=breaks, scales=list(draw=TRUE, tck=c(1,0)), col="grey26", lwd=0.5))
    # using 'sf' package
    print(z_plot)
    }

}
